src/promptflow-evals/promptflow/evals/evaluate/_evaluate.py

```diff
@@ -59,12 +70,95 @@ def _validation(target, data, evaluators, output_path, tracking_uri, evaluation_
             raise ValueError("evaluation_name must be a string.")
 
     try:
-        data_df = pd.read_json(data, lines=True)
+        initial_data_df = pd.read_json(data, lines=True)
     except Exception as e:
-        raise ValueError(f"Failed to load data from {data}. Please validate it is a valid jsonl data. Error: {str(e)}.")
+        raise ValueError(
+            f"Failed to load data from {data}. Please validate it is a valid jsonl data. Error: {str(e)}.")
 
-    for evaluator_name, evaluator in evaluators.items():
-        _validate_input_data_for_evaluator(evaluator, evaluator_name, data_df)
+    _validate_columns(initial_data_df, evaluators, target)
+    return initial_data_df
+
+
+def _validate_columns(df: pd.DataFrame, evaluators: Dict[str, Any], target: Optional[Callable]) -> None:
+    """
+    Check that all columns needed by evaluator or target function are present.
+
+    :keyword df: The data frame to be validated.
+    :paramtype df: pd.DataFrame
+    :keyword evaluators: The dictionary of evaluators.
+    :paramtype evaluators: Dict[str, Any]
+    :keyword target: The callable to be applied to data set.
+    :paramtype target: Optional[Callable]
+    """
+    if target:
+        # If the target function is given, it may return
+        # several columns and hence we cannot check the availability of columns
+        # without knowing target function semantics.
+        # Instead, here we will validate the columns, taken by target.
+        _validate_input_data_for_evaluator(target, None, df, is_target_fn=True)
+    else:
+        for evaluator_name, evaluator in evaluators.items():
+            _validate_input_data_for_evaluator(evaluator, evaluator_name, df)
+
+
+def _apply_target_to_data(
+        target: Callable,
+        data: str,
+        pf_client: PFClient,
+        initial_data: pd.DataFrame) -> Tuple[str, pd.DataFrame, List[str]]:
+    """
+    Apply the target function to the data set and save the data to temporary file.
+
+    :keyword target: The function to be applied to data.
+    :paramtype target: Callable
+    :keyword data: The path to input jsonl file.
+    :paramtype data: str
+    :keyword pf_client: The promptflow client to be used.
+    :paramtype pf_client: PFClient
+    :keyword initial_data: The data frame with the loaded data.
+    :paramtype initial_data: pd.DataFrame
+    :return: The tuple, containing path, data frame and the list of added columns.
+    :rtype: Tuple[str, pd.DataFrame, List[str]]
+    """
+    # We are manually creating the temporary directory for the flow
+    # because the way tempdir remove temporary directories will
+    # hang the debugger, because promptflow will keep flow directory.
+    saved_flow = f'flow_{uuid.uuid1()}'
+    os.makedirs(saved_flow)

```

Comment:
We should be able to pass the target function directly to pf.run without needing to save it to disk

Reply:
That will work; however it will fail if we call the evaluate from the Jupyter notebook (however, it will pass the test).
This will happen ecause of [this](https://github.com/microsoft/promptflow/blob/4af2968d24279028d16e552a247f1073cdf1efed/src/promptflow-core/promptflow/_core/tool_meta_generator.py#L125) line. The notebook's module will be marked as __main__, however, if will not have the function, defined in one of the cells.

Reply:
Created a bug for promptflow team and using function for now. https://msdata.visualstudio.com/Vienna/_workitems/edit/3144908